src/lib.rs

```diff
@@ -0,0 +1,324 @@
+            let key_size = key.as_bytes().len() * mem::size_of::<*const c_char>();
+            unsafe {
+                randomx_init_cache(result.cache, key_ptr, key_size);
+            }
+            Ok(result)
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct RandomXDataset {
+    dataset: *mut randomx_dataset,
+    dataset_start: c_ulong,
+    dataset_count: c_ulong,
+}
+
+impl Drop for RandomXDataset {
+    fn drop(&mut self) {
+        unsafe {
+            randomx_release_dataset(self.dataset);
+        }
+    }
+}
+
+impl RandomXDataset {
+    pub fn new(
+        flags: Vec<RandomXFlag>,
+        cache: &RandomXCache,
+        start: c_ulong,
+    ) -> Result<RandomXDataset, CreationError> {
+        let count = c_ulong::from(RANDOMX_DATASET_ITEM_SIZE - 1) - start;
+
+        let mut flag: c_uint = RandomXFlag::FlagDefault.value();
+        for f in flags {
+            if f == RandomXFlag::FlagLargePages {
+                flag |= f.value();
+            }
+        }
+        let test = unsafe { randomx_alloc_dataset(flag) };
+        if test.is_null() {
+            Err(CreationError)
+        } else {
+            let result = RandomXDataset {
+                dataset: test,
+                dataset_start: start,
+                dataset_count: count,
+            };
+            unsafe {
+                randomx_init_dataset(result.dataset, cache.cache, start, count);
+            }
+            Ok(result)
+        }
+    }
+
+    pub fn count(&self) -> c_ulong {
+        unsafe { randomx_dataset_item_count() }
+    }
+
+    pub fn get_data(&self) -> Vec<u8> {
+        let memory = unsafe { randomx_get_dataset_memory(self.dataset) };
+        let mut result = Vec::new();
+        for i in self.dataset_start..self.dataset_count {
+            result.push(unsafe { memory.offset(i as isize) } as u8);
+        }
+        result
+    }
+}
+
+#[derive(Debug)]
+pub struct RandomXVM {
+    vm: *mut randomx_vm,
+}
+
+impl Drop for RandomXVM {
+    fn drop(&mut self) {
+        unsafe {
+            randomx_destroy_vm(self.vm);
+        }
+    }
+}
+
+impl RandomXVM {
+    pub fn new(
+        flags: Vec<RandomXFlag>,
+        cache: &RandomXCache,
+        dataset: Option<&RandomXDataset>,
+    ) -> Result<RandomXVM, CreationError> {
+        let mut flag: c_uint = RandomXFlag::FlagDefault.value();
+        for f in flags {
+            flag |= f.value();
+        }
+        let test: *mut randomx_vm;
+        match dataset {
+            Some(data) => unsafe { test = randomx_create_vm(flag, cache.cache, data.dataset) },
+            None => unsafe { test = randomx_create_vm(flag, cache.cache, ptr::null_mut()) },
+        }
+        if test.is_null() {
+            return Err(CreationError);
+        }
+        let result = RandomXVM { vm: test };
+        Ok(result)
+    }
+
+    pub fn reinit_cache(&self, cache: &RandomXCache) {
+        unsafe {
+            randomx_vm_set_cache(self.vm, cache.cache);
+        }
```

Comment:
Is is possible for `cache` or it's fields to be undefined / null here? 
If so, are there checks we can make to satisfy safety guarantees?